src/documents/templates/ExtractItem.ts

```diff
@@ -0,0 +1,113 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from "vscode";
+import { IActionContext, IAzureUserInput } from "vscode-azureextensionui";
+import { ObjectValue } from "../../language/json/JSON";
+import { DeploymentTemplateDoc } from "./DeploymentTemplateDoc";
+import { InsertItem } from "./insertItem";
+export class ExtractItem {
+    // tslint:disable-next-line:no-empty
+    constructor(private ui: IAzureUserInput) {
+    }
+
+    public async extractParameter(editor: vscode.TextEditor, template: DeploymentTemplateDoc, context: IActionContext): Promise<void> {
+        let selection = this.expandSelection(editor.selection, editor.document, template, editor);
```

Comment:
I think there's no guarantee that the editor contents haven't changed since this command was fired.  So it might be possible to get unexpected results.  The easiest solution is to pass the current template as an argument to the extract command and if it's not the same, simply abort.  You could also pass in precalculated range/etc in through the commands if you want.

Reply:
Are you suggesting that I send in the uri of the current document as an argument and abort if they are not the same?
Or somehow get the context of the current document?

Reply:
Rethinking this, since this is vscode controlled on both sides, it's probably safe.